### PR TITLE
Improve dependency tests

### LINKED ISSUE #30332: [Issue] Improve dependency tests
This issue is automatically created based on existing pull request: magento/magento2#30331: Improve dependency tests

---------
### Description (*)
Resolve issues of running Dependency Tests for not Github installation.
Add Tests Coverage for \Magento\TestFramework\Dependency\Route\RouteMapper.
Fix failed tests when Route Id and Front Name can be different in routes.xml

### PR DESCRIPTION
**Pull request by swnsma**
### Description (*)
Resolve issues of running Dependency Tests for not Github installation.
Add Tests Coverage for \Magento\TestFramework\Dependency\Route\RouteMapper.
Fix failed tests when Route Id and Front Name can be different in routes.xml

### Resolved issues:
1. [x] resolves magento/magento2#30332: Improve dependency tests

### Related Pull Requests
1. https://github.com/magento/security-package/pull/278

### CODE REVIEW

**Comment by swnsma** [comment]
@magento create issue

**Comment by swnsma** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@swnsma could you add more details for your changes?

File: dev/tests/static/testsuite/Magento/Test/Integrity/Dependency/DeclarativeSchemaDependencyProvider.php [hunk 1]
```diff
@@ -214,9 +213,10 @@ private function getDeclarativeSchema(): array
...
                 $moduleDeclaration = array_replace_recursive(
                     $moduleDeclaration,
-                    [self::SCHEMA_ENTITY_TABLE => [
-                        $tableName => $tableDeclaration,
-                    ]
+                    [
+                        self::SCHEMA_ENTITY_TABLE => [
+                                $tableName => $tableDeclaration,
+                            ]
                     ]
```
Comment (ihor-sviziev) [inline dev/tests/static/testsuite/Magento/Test/Integrity/Dependency/DeclarativeSchemaDependencyProvider.php]:
```suggestion
                    [
                        self::SCHEMA_ENTITY_TABLE => [
                            $tableName => $tableDeclaration,
                        ]
                    ]
```

File: dev/tests/static/testsuite/Magento/Test/Integrity/DependencyTest.php [hunk 2]
```diff
@@ -342,20 +318,29 @@ private static function getAppCodeDir()
...
+
+        /** @var \Magento\Framework\Component\ComponentFile[] $configs */
+        $configs = Files::init()->getConfigFiles(
+            'db_schema_whitelist.json',
+            [],
+            false,
+            true
+        );
+        foreach ($configs as $file) {
+            $dbSchemaWhitelist = (array)json_decode(file_get_contents($file->getFullPath()));
```
Comment (ihor-sviziev) [inline dev/tests/static/testsuite/Magento/Test/Integrity/DependencyTest.php:361]: https://www.php.net/manual/en/function.json-decode.php
image

```suggestion
            $dbSchemaWhitelist = json_decode(file_get_contents($file->getFullPath()), true);
```

**Comment by swnsma** [comment]
Hi @ihor-sviziev

1. Change searching of module folder and files from 'app/code' to it's real location for Route Mapper.
2. Change searching module name based on its path (like app/code/Magento/Module) to it's real name, independently from the path for DeclarativeSchemaDependencyProvider.
3. Add tests coverage for Route Mapper.
4. Fix route mapper to search dependency by route front name (for `getUrl` method it does not matter if you pass front name or router id).
5. Fix Dependency Test: now it starts to collect DB configs from modules folder instead of using pattern (app/code/*/*/etc/db_schema_whitelist.json).

File: dev/tests/static/testsuite/Magento/Test/Integrity/DependencyTest.php [hunk 3]
```diff
@@ -374,15 +359,24 @@ private static function getTableToPrimaryModuleMap(): array
...
+        /** @var \Magento\Framework\Component\ComponentFile[] $configs */
+        $configs = Files::init()->getConfigFiles(
+            'db_schema_whitelist.json',
+            [],
+            false,
+            true
+        );
+
+        foreach ($configs as $file) {
+            $dbSchemaWhitelist = (array)json_decode(file_get_contents($file->getFullPath()));
```
Comment (ihor-sviziev) [inline dev/tests/static/testsuite/Magento/Test/Integrity/DependencyTest.php:402]:
```suggestion
            $dbSchemaWhitelist = json_decode(file_get_contents($file->getFullPath()), true);
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @swnsma ,
Static tests are failing, could you check why?

**Comment by swnsma** [comment]
@magento run Static Tests

**Comment by swnsma** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
@swnsma tests still failing, please fix them

**Comment by swnsma** [comment]
Hi @ihor-sviziev,
Fix for failed test is in security repo PR (please see related PRs).
So I can not fix in in current repository.

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@sidolov @slavvka could you help with running tests with related Pull Request? Seems like it doesn't work by now.

**Comment by ihor-sviziev** [comment]
@magento run WebAPI Tests, Functional Tests B2B, Functional Tests EE, Functional Tests CE

**Comment by sidolov** [comment]
@ihor-sviziev it was an issue with PR format, I have changed it, and builds should work

**Comment by ihor-sviziev** [comment]
@magento run Static Tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests CE, Functional Tests B2B

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests CE

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8330](https://jira.corp.magento.com/browse/ENGCOM-8330) has been created to process this Pull Request

**Comment by engcom-Bravo** [comment]
QA not applicable

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @swnsma ,
Could you resolve conflict?

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by swnsma** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B

**Comment by swnsma** [comment]
@magento run Functional Tests B2B

**Comment by ihor-sviziev** [comment]
@engcom-Foxtrot tests are passing now ✔

**Comment by swnsma** [comment]
Hi @engcom-Foxtrot,

Is there anything else should be done in this PR?

**Comment by swnsma** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@swnsma I think this PR is stuck because we still have not processed higher priority PRs

**Comment by ihor-sviziev** [comment]
@magento run Database Compare, Functional Tests EE, Static Tests, WebAPI Tests

**Comment by ihor-sviziev** [comment]
@sidolov @gabrieldagama could you tell why the static tests are failing?

**Comment by swnsma** [comment]
@magento run  Functional Tests EE, Static Tests, WebAPI Tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
@swnsma, it looks like the dependency test you were fixing is failing, so behavior has changed. Please double-check that.

**Comment by swnsma** [comment]
@magento run all tests

**Comment by swnsma** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by swnsma** [comment]
@magento run all tests

**Comment by engcom-Dash** [comment]
Hello @swnsma,

Thank you for your contributions!

Please resolve the conflict so we can proceed with this PR.

Thanks again!

**Comment by engcom-Dash** [comment]
Closing this PR since it has not been updated in a while. Please feel free to reopen if needed.

**Pull request closed without merging**